src/query/src/range_select/plan_rewrite.rs

```diff
@@ -0,0 +1,424 @@
+                    ));
+                } else {
+                    self.by = by.clone();
+                }
+                range_fn.expr = self.gen_range_expr(func_name, args)?;
+                let alias = Expr::Column(Column::from_name(range_fn.expr.display_name()?));
+                self.range_fn.push(range_fn);
+                return Ok(alias);
+            }
+        }
+        Ok(node)
+    }
+}
+
+pub struct RangePlanRewriter {
+    table_provider: DfTableSourceProvider,
+    context_provider: DfContextProviderAdapter,
+}
+
+impl RangePlanRewriter {
+    pub fn new(
+        table_provider: DfTableSourceProvider,
+        context_provider: DfContextProviderAdapter,
+    ) -> Self {
+        Self {
+            table_provider,
+            context_provider,
+        }
+    }
+
+    pub async fn rewrite(&mut self, plan: LogicalPlan) -> Result<LogicalPlan> {
+        match self.rewrite_logical_plan(&plan).await? {
+            Some(new_plan) => Ok(new_plan),
+            None => Ok(plan),
+        }
+    }
+
+    #[async_recursion]
+    async fn rewrite_logical_plan(&mut self, plan: &LogicalPlan) -> Result<Option<LogicalPlan>> {
+        let inputs = plan.inputs();
+        let mut new_inputs = Vec::with_capacity(inputs.len());
+        for input in &inputs {
+            new_inputs.push(self.rewrite_logical_plan(input).await?)
+        }
+        match plan {
+            LogicalPlan::Projection(Projection { expr, input, .. })
+                if have_range_in_exprs(expr) =>
+            {
+                let input = if let Some(new_input) = new_inputs[0].take() {
+                    Arc::new(new_input)
+                } else {
+                    input.clone()
+                };
+                let (time_index, default_by) = self.get_index_by(input.schema().clone()).await?;
+                let mut range_rewriter = RangeRewriter {
+                    align: Duration::default(),
+                    by: vec![],
+                    range_fn: vec![],
+                    context_provider: &self.context_provider,
+                };
+                let new_expr = expr
+                    .iter()
+                    .map(|expr| expr.clone().rewrite(&mut range_rewriter))
+                    .collect::<DFResult<Vec<_>>>()
+                    .context(DataFusionSnafu)?;
+                if range_rewriter.by.is_empty() {
+                    range_rewriter.by = default_by;
+                }
+                let range_plan = RangeSelect::try_new(
+                    input.clone(),
+                    range_rewriter.range_fn,
+                    range_rewriter.align,
+                    time_index,
+                    range_rewriter.by,
+                )?;
+                let project_plan = LogicalPlanBuilder::from(LogicalPlan::Extension(Extension {
+                    node: Arc::new(range_plan),
+                }))
+                .project(new_expr)
```

Comment:
looks like we don't need an extra projection? like aggregate/window itself can be used as the final result

Reply:
For select like `SELECT (avg(field_0)+sum(field_1))/4 ....`,  we need range plan to gererate schema like `#avg(field_0), #sum(field_1)`, and use additional projection plan to dispose `(#avg(field_0)+#sum(field_1))/4`, expr rewriter write `avg(field_0)` to a plain column, so we can read value from range plan's schema

Reply:
I don't get it. Complex expr like `SELECT (avg(field_0)+sum(field_1))/4 RANGE '5m'` would only result in one column in the range plan's output schema, and this column's name is something like `(#avg(field_0)+#sum(field_1))/4` (follow the conversion). If you only want to change the column's name, it can be done inside the range plan

Reply:
No, `avg(field_0)` and `sum(field_1)` will be two column in range plan's output schema, `avg(field_0)` and `sum(field_1)` are the column names. And we leave complex expr  `(avg(field_0)+sum(field_1))/4` to projection plan,   projection plan find  `avg(field_0)` and `sum(field_1)` from range plan and do rest caculation

Reply:
the `#` just a mark. `#avg(field_0)`  just a column with name `avg(field_0)`,  `avg(field_0)` mean `avg` on `field_0`

Reply:
I misunderstand that `(avg(field_0)+sum(field_1))/4 RANGE '5m'` are two range expressions. In this case we do need a projection to calculate `+` and `/ 4`. But only when there is post-calculation that we need an extra Projection. Not unconditional